File: sliding_window/904_fruit_into_baskets.go
func totalFruit(fruits []int) int {
	begin := 0
	// тут храним фрукты и их количество
	windowState := make(map[int]int)
	result := 0

	for end := range len(fruits) {
		windowState[fruits[end]]++

		// если фруктов стало больше чем надо
		for len(windowState) > 2 {
			// вычитаем один фрукт (он точно будет т.к. end быстрее begin)
			windowState[fruits[begin]]--
			// если счетчик ноль, удаляем из мапы
			if windowState[fruits[begin]] == 0 {
				delete(windowState, fruits[begin])
			}
			// сдвигаем рамку каждый раз
			begin++
		}
		result = max(result, end-begin+1)
	}
	return result
}

File: sliding_window/904_fruit_into_baskets.py
from collections import defaultdict

def totalFruit(fruits: list[int]) -> int:
	begin = 0 
	window_state = defaultdict(int)
	result = 0

	for end in range(len(fruits)):
		window_state[fruits[end]] += 1
		while len(window_state) > 2:
			window_state[fruits[begin]] -= 1
			if window_state[fruits[begin]] == 0:
				del window_state[fruits[begin]]
			begin += 1
			
		result = max(result, end - begin + 1)
	
	return result
